File: src/main/java/org/willxu/algorithm/service/other/MergeTwoSortedLists.java
package org.willxu.algorithm.service.other;

import org.willxu.algorithm.provide.ListNode;

public interface MergeTwoSortedLists {
    /**
     * Merge two sorted linked lists and return it as a sorted list.
     * The list should be mage by splicing together the nodes of the
     * first two lists.
     *
     * @param l1 sorted linked lists
     * @param l2 sorted linked lists
     * @return sorted linked lists merged by l1 and l2
     */
    ListNode mergeTwoLists(ListNode l1, ListNode l2);
}

File: src/leetcode/p0000_0099/p0021_merge_two_sorted_lists/merge_two_sorted_lists.py
from abc import ABC, abstractmethod
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class MergeTwoSortedLists(ABC):
    """
    You are given the heads of two sorted linked lists list1 and list2.

    Merge the two lists into one sorted list. The list should be made by splicing together the nodes of the first two lists.

    Return the head of the merged linked list.
    """
    @abstractmethod
    def merge_two_lists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        """
        The number of nodes in both lists is in the range [0, 50].
        -100 <= Node.val <= 100
        Both list1 and list2 are sorted in non-decreasing order.
        """
        pass

class MergeTwoSortedListsRecursive(MergeTwoSortedLists):
    def merge_two_lists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1:
            return list2
        if not list2:
            return list1
        if list1.val <= list2.val:
            list1.next = self.merge_two_lists(list1.next, l2)
            return list1
        else:
            list2.next = self.merge_two_lists(list1, list2.next)
            return list2
        return ListNode(0)
